experiment/nursery/diskcache/cache.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// cache implements disk backed cache storage for use in nursery
+package diskcache
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ReadHandler should be implemeted by cache users for use with Cache.Get
+type ReadHandler func(exists bool, contents io.ReadSeeker) error
+
+// Cache implements disk backed cache storage
+type Cache struct {
+	diskRoot string
+}
+
+// NewCache returns a new Cache given the root directory that should be used
+// on disk for cache storage
+func NewCache(diskRoot string) *Cache {
+	return &Cache{
+		diskRoot,
+	}
+}
+
+func (c *Cache) keyToPath(key string) string {
+	return filepath.Join(c.diskRoot, key)
+}
+
+// DiskRoot returns the root directory containing all on-disk cache entries
+func (c *Cache) DiskRoot() string {
+	return c.diskRoot
+}
+
+// file path helper
+func exists(path string) bool {
+	_, err := os.Stat(path)
+	return !os.IsNotExist(err)
+}
+
+// file path helper
+func ensureDir(dir string) {
+	if exists(dir) {
+		return
+	}
+	log.WithError(os.MkdirAll(dir, os.FileMode(0744))).Infof("MkDirAll(%s)", dir)
+}
+
+// Put copies the content reader until the end into the cache at key
+// if contentSHA256 is not "" then the contents will only be stored in the
+// cache if the content's hex string SHA256 matches
+func (c *Cache) Put(key string, content io.Reader, contentSHA256 string) error {
+	// make sure directory exists
+	path := c.keyToPath(key)
+	dir := filepath.Dir(path)
+	ensureDir(dir)
+
+	// create a temp file to get the content on disk
+	temp, err := ioutil.TempFile(dir, "temp-put")
+	if err != nil {
+		return fmt.Errorf("failed to create cache entry: %v", err)
+	}
+
+	// fast path copying when not hashing content,s
+	if contentSHA256 == "" {
+		io.Copy(temp, content)
+		if err != nil {
+			os.Remove(temp.Name())
```

Comment:
check error, and for all `os.Remove` below

Reply:
done